managers/memory/handle_mq.c

```diff
@@ -0,0 +1,301 @@
+
+	strcpy(msg_data, item->msg_data);
+	*msg_size = item->msg_size;
+
+	list_del(&item->list);
+
+	kfree(item->msg_data);	
+	kfree(item);
+	name_nid_dict = next;
+
+	return MSG_RET_SUCCESS;
+}
+
+
+/*
+ * print api, for debug
+ */ 
+void print(struct list_head* name_nid_dict){
+	struct mc_msg_queue *pos = NULL;
+	list_for_each_entry(pos, name_nid_dict, list){
+		printk(pos->msg_data);
+		printk(" ");
+		printk("message data: %d\n", pos->msg_size);
+	}
+}
+
+void free_all(struct list_head* name_nid_dict){
+
+	struct list_head* cur=name_nid_dict->next;
+	while(cur != name_nid_dict){		
+		struct mc_msg_queue* item = list_entry(cur, struct mc_msg_queue, list);
+
+		struct list_head* tmp = cur->next;
+		list_del(cur);
+
+		cur = tmp;
+		kfree(item->msg_data);
+		kfree(item);
+	}
+}
+
+/*
+ * api for the message queue data structure in mComponent kernel side
+ */ 
+unsigned int mc_mq_open(char* mq_name, unsigned int max_size)
+{
+
+/*
+ * what if we already got a message queue with that name in the name map?	
+ */
+	unsigned long flags;
+	spin_lock_irqsave(&map_lock, flags);
+
+	struct name_mq_map *pos, *target = NULL;
+	list_for_each_entry(pos, &addr_map, list){
+
+		if(strcmp(mq_name, pos->mq_name)==0){
+			target = pos;		
+		}
+	}
+
+	if(target != NULL){
+		spin_unlock_irqrestore(&map_lock, flags);
+		return MSG_RET_FAIL;
+	}
+	
+	spin_unlock_irqrestore(&map_lock, flags);
+
+	struct name_mq_map* tmp;
+
+	tmp = kmalloc(sizeof(struct name_mq_map), GFP_KERNEL);
+	if(!tmp){
+		printk("allocated wrong\n");
+		return MSG_RET_FAIL;
+	}
+
+	strcpy(tmp->mq_name, mq_name);
+	tmp->max_size = max_size;
+	tmp->mq = kmalloc(sizeof(struct list_head), GFP_KERNEL);
+
+	INIT_LIST_HEAD(tmp->mq);	
+
+	/* init the lock */
+	spin_lock_init(&tmp->mq_lock);
+	spin_lock_irqsave(&map_lock, flags);
+
+	list_add_tail(&tmp->list, &addr_map);	
+
+	spin_unlock_irqrestore(&map_lock, flags);
+
+	kfree(tmp);
+	return MSG_RET_SUCCESS;
+}
+
+unsigned int mc_mq_close(char* mq_name){
+	unsigned long flags;
+	spin_lock_irqsave(&map_lock, flags);
+
+	/* make sure the message queue exists*/
+	struct name_mq_map *pos, *target = NULL;
+	list_for_each_entry(pos, &addr_map, list){
+
+		if(strcmp(mq_name, pos->mq_name)==0){
+			target = pos;		
+		}
+	}
+
+	if(target ==NULL){
+		spin_unlock_irqrestore(&map_lock, flags);	
+		return MSG_RET_FAIL;	
+	}
+
+	list_del(&target->list);
+
+	spin_unlock_irqrestore(&map_lock, flags);
+
+	/* possible leak
+         * what if message queue still gots some thing, then we free then
+         */
+	free_all(target->mq);
+	kfree(target->mq);
+	kfree(target);
+
+	return MSG_RET_SUCCESS;
+}
+
+
+unsigned int mc_mq_send(char *mq_name, char* msg_data, unsigned int msg_size){
+
+	/* find out where is our mq head pointer */
+	struct name_mq_map *pos, *target = NULL;
+	list_for_each_entry(pos, &addr_map, list){
+		if(strcmp(mq_name, pos->mq_name)==0){
+			target = pos;		
+		}
+	}
+	if(target != NULL){
+		/* spin lock acquire here */
+		unsigned long flags;
+
```

Comment:
what if the mq get closed after your list search and before your append?

Reply:
cool, didn't expect that, I add the synchronization